rust/chains/hyperlane-cosmos/src/providers/mod.rs

```diff
@@ -116,6 +116,55 @@ impl CosmosProvider {
 
         Ok((account_id, signer_info.sequence))
     }
+
+    /// Calculates the sender and the nonce for the transaction.
+    /// We use `payer` of the fees as the sender of the transaction, and we search for `payer`
+    /// signature information to find the nonce.
+    /// If `payer` is not specified, we use the account which signed the transaction first, as
+    /// the sender.
+    fn sender_and_nonce(&self, tx: &Tx) -> Result<(H256, SequenceNumber), ChainCommunicationError> {
+        let (sender, nonce) = tx
+            .auth_info
+            .fee
+            .payer
+            .as_ref()
+            .map(|payer| self.search_payer_in_signer_infos(&tx.auth_info.signer_infos, payer))
+            .map_or_else(
+                || {
+                    let signer_info = tx.auth_info.signer_infos.get(0).ok_or_else(|| {
+                        ChainCommunicationError::from_other_str("no signer info in default signer")
+                    })?;
+                    self.convert_signer_info_into_account_id_and_nonce(signer_info)
+                },
+                |p| p,
+            )
+            .map(|(a, n)| CosmosAddress::from_account_id(a).map(|a| (a.digest(), n)))??;
+        Ok((sender, nonce))
+    }
+
+    /// Extract contract address from transaction.
+    /// Assumes that there is only one `MsgExecuteContract` message in the transaction
```

Comment:
When would there not be just one `MsgExecuteContract` message in a transaction? Thinking that we'll eventually batch submissions in the same tx and not sure if that affects this assumption - but this can be handled at that time

Reply:
Yes, the challenge is several messages which execute different contracts. Current implementation does not identify the type of contract. The message which is sent to contract (json) contains only method and arguments, but it does not contain the type of contract. So, if we have several messages, we need to filter out execution of unrelated contracts as well. There is a lot to think about.